app/models/goods_nomenclatures/nested_set.rb

```diff
@@ -0,0 +1,176 @@
+          .with_validity_dates(:descendant_nodes)
+          .select_append(:descendant_nodes__depth)
+          .join(Sequel.as(:goods_nomenclature_tree_nodes, :origin_nodes)) do |origin_table, descendants_table, _join_clauses|
+            descendants_position = Sequel.qualify(descendants_table, :position)
+            descendants_depth    = Sequel.qualify(descendants_table, :depth)
+            origin_position      = Sequel.qualify(origin_table, :position)
+            origin_depth         = Sequel.qualify(origin_table, :depth)
+
+            (descendants_depth > origin_depth) &
+              (descendants_position > origin_position) &
+              model.validity_dates_filter(origin_table) &
+              (descendants_position <
+                Sequel.function(
+                  :coalesce,
+                  TreeNode.next_sibling(origin_position, origin_depth),
+                  END_OF_TREE,
+                )
+              )
+          end
+      end
+
+      many_to_many :ns_children,
+                   left_primary_key: :goods_nomenclature_sid,
+                   left_key: Sequel.qualify(:origin_nodes, :goods_nomenclature_sid),
+                   right_primary_key: :goods_nomenclature_sid,
+                   right_key: :goods_nomenclature_sid,
+                   class_name: '::GoodsNomenclature',
+                   join_table: Sequel.as(:goods_nomenclature_tree_nodes, :child_nodes),
+                   read_only: true do |ds|
+        ds.order(:child_nodes__position)
+          .with_validity_dates(:child_nodes)
+          .select_append(:child_nodes__depth)
+          .join(Sequel.as(:goods_nomenclature_tree_nodes, :origin_nodes)) do |origin_table, descendants_table, _join_clauses|
+            descendants_position = Sequel.qualify(descendants_table, :position)
+            descendants_depth    = Sequel.qualify(descendants_table, :depth)
+            origin_position      = Sequel.qualify(origin_table, :position)
+            origin_depth         = Sequel.qualify(origin_table, :depth)
+
+            (descendants_depth =~ (origin_depth + 1)) &
+              (descendants_position > origin_position) &
+              model.validity_dates_filter(origin_table) &
+              (descendants_position <
+                Sequel.function(
+                  :coalesce,
+                  TreeNode.next_sibling(origin_position, origin_depth),
+                  END_OF_TREE,
+                )
+              )
+          end
+      end
+    end
+
+    def recursive_ancestor_populator(ancestors)
+      @associations ||= { ns_ancestors: ancestors }
+
+      parents_ancestors = ancestors.dup
+      parent = parents_ancestors.pop
+      @associations[:ns_parent] ||= parent
+      return if ancestors.empty?
+
+      parent.recursive_ancestor_populator(parents_ancestors)
+    end
+
+    def recursive_descendant_populator(descendants, parent = nil)
+      @associations ||= { ns_descendants: descendants }
```

Comment:
Is this populating the same association cache that sequel uses internally?

Reply:
Yes